### ISSUE TITLE
Devicemapper stalling

### ISSUE THREAD

**Issue by Calpicow** [moby/moby#30003]
**Description**

Devicemapper stalls which causes Docker to stop working completely.

[dump](https://gist.github.com/Calpicow/e9ba0de603949c2894b749239a697eb2)
[dmesg](https://gist.github.com/Calpicow/6cae67d065b83b8372824d90d168d109)
[dmsetup](https://gist.github.com/Calpicow/b739a88278e16600319be2a5e6379ad5)

```
Jan  6 01:41:19 ip-100-64-32-70 kernel: INFO: task kworker/u31:1:91 blocked for more than 120 seconds.
Jan  6 01:41:19 ip-100-64-32-70 kernel: "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
Jan  6 01:41:19 ip-100-64-32-70 kernel: kworker/u31:1   D ffff880201fc98e0     0    91      2 0x00000000
Jan  6 01:41:19 ip-100-64-32-70 kernel: Workqueue: kdmremove do_deferred_remove [dm_mod]
Jan  6 01:41:20 ip-100-64-32-70 kernel: ffff88020141bcf0 0000000000000046 ffff8802044ce780 ffff88020141bfd8
Jan  6 01:41:20 ip-100-64-32-70 kernel: ffff88020141bfd8 ffff88020141bfd8 ffff8802044ce780 ffff880201fc98d8
Jan  6 01:41:20 ip-100-64-32-70 kernel: ffff880201fc98dc ffff8802044ce780 00000000ffffffff ffff880201fc98e0
Jan  6 01:41:20 ip-100-64-32-70 kernel: Call Trace:
Jan  6 01:41:20 ip-100-64-32-70 kernel: [<ffffffff8163b959>] schedule_preempt_disabled+0x29/0x70
Jan  6 01:41:20 ip-100-64-32-70 kernel: [<ffffffff81639655>] __mutex_lock_slowpath+0xc5/0x1c0
Jan  6 01:41:20 ip-100-64-32-70 kernel: [<ffffffff81638abf>] mutex_lock+0x1f/0x2f
Jan  6 01:41:20 ip-100-64-32-70 kernel: [<ffffffffa0392e9d>] __dm_destroy+0xad/0x340 [dm_mod]
Jan  6 01:41:20 ip-100-64-32-70 kernel: [<ffffffffa03947e3>] dm_destroy+0x13/0x20 [dm_mod]
Jan  6 01:41:20 ip-100-64-32-70 kernel: [<ffffffffa0398d6d>] dm_hash_remove_all+0x6d/0x130 [dm_mod]
Jan  6 01:41:20 ip-100-64-32-70 kernel: [<ffffffffa039b50a>] dm_deferred_remove+0x1a/0x20 [dm_mod]
Jan  6 01:41:20 ip-100-64-32-70 kernel: [<ffffffffa0390dae>] do_deferred_remove+0xe/0x10 [dm_mod]
Jan  6 01:41:20 ip-100-64-32-70 kernel: [<ffffffff8109d5fb>] process_one_work+0x17b/0x470
Jan  6 01:41:20 ip-100-64-32-70 kernel: [<ffffffff8109e3cb>] worker_thread+0x11b/0x400
Jan  6 01:41:20 ip-100-64-32-70 kernel: [<ffffffff8109e2b0>] ? rescuer_thread+0x400/0x400
Jan  6 01:41:20 ip-100-64-32-70 kernel: [<ffffffff810a5aef>] kthread+0xcf/0xe0
Jan  6 01:41:20 ip-100-64-32-70 kernel: [<ffffffff810a5a20>] ? kthread_create_on_node+0x140/0x140
Jan  6 01:41:20 ip-100-64-32-70 kernel: [<ffffffff81645818>] ret_from_fork+0x58/0x90
Jan  6 01:41:20 ip-100-64-32-70 kernel: [<ffffffff810a5a20>] ? kthread_create_on_node+0x140/0x140
Jan  6 01:41:20 ip-100-64-32-70 kernel: INFO: task dockerd:31587 blocked for more than 120 seconds.
Jan  6 01:41:20 ip-100-64-32-70 kernel: "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
Jan  6 01:41:20 ip-100-64-32-70 kernel: dockerd         D 0000000000000000     0 31587      1 0x00000080
Jan  6 01:41:20 ip-100-64-32-70 kernel: ffff8800e768fab0 0000000000000086 ffff880034215c00 ffff8800e768ffd8
Jan  6 01:41:20 ip-100-64-32-70 kernel: ffff8800e768ffd8 ffff8800e768ffd8 ffff880034215c00 ffff8800e768fbf0
Jan  6 01:41:20 ip-100-64-32-70 kernel: ffff8800e768fbf8 7fffffffffffffff ffff880034215c00 0000000000000000
Jan  6 01:41:20 ip-100-64-32-70 kernel: Call Trace:
Jan  6 01:41:20 ip-100-64-32-70 kernel: [<ffffffff8163a879>] schedule+0x29/0x70
Jan  6 01:41:20 ip-100-64-32-70 kernel: [<ffffffff81638569>] schedule_timeout+0x209/0x2d0
Jan  6 01:41:20 ip-100-64-32-70 kernel: [<ffffffff8108e4cd>] ? mod_timer+0x11d/0x240
Jan  6 01:41:20 ip-100-64-32-70 kernel: [<ffffffff8163ac46>] wait_for_completion+0x116/0x170
Jan  6 01:41:20 ip-100-64-32-70 kernel: [<ffffffff810b8c10>] ? wake_up_state+0x20/0x20
Jan  6 01:41:20 ip-100-64-32-70 kernel: [<ffffffff810ab676>] __synchronize_srcu+0x106/0x1a0
Jan  6 01:41:20 ip-100-64-32-70 kernel: [<ffffffff810ab190>] ? call_srcu+0x70/0x70
Jan  6 01:41:20 ip-100-64-32-70 kernel: [<ffffffff81219e3f>] ? __sync_blockdev+0x1f/0x40
Jan  6 01:41:20 ip-100-64-32-70 kernel: [<ffffffff810ab72d>] synchronize_srcu+0x1d/0x20
Jan  6 01:41:20 ip-100-64-32-70 kernel: [<ffffffffa039318d>] __dm_suspend+0x5d/0x220 [dm_mod]
Jan  6 01:41:20 ip-100-64-32-70 kernel: [<ffffffffa0394c9a>] dm_suspend+0xca/0xf0 [dm_mod]
Jan  6 01:41:20 ip-100-64-32-70 kernel: [<ffffffffa0399fe0>] ? table_load+0x380/0x380 [dm_mod]
Jan  6 01:41:20 ip-100-64-32-70 kernel: [<ffffffffa039a174>] dev_suspend+0x194/0x250 [dm_mod]
Jan  6 01:41:20 ip-100-64-32-70 kernel: [<ffffffffa0399fe0>] ? table_load+0x380/0x380 [dm_mod]
Jan  6 01:41:20 ip-100-64-32-70 kernel: [<ffffffffa039aa25>] ctl_ioctl+0x255/0x500 [dm_mod]
Jan  6 01:41:20 ip-100-64-32-70 kernel: [<ffffffff8112482d>] ? call_rcu_sched+0x1d/0x20
Jan  6 01:41:20 ip-100-64-32-70 kernel: [<ffffffffa039ace3>] dm_ctl_ioctl+0x13/0x20 [dm_mod]
Jan  6 01:41:20 ip-100-64-32-70 kernel: [<ffffffff811f1e75>] do_vfs_ioctl+0x2e5/0x4c0
Jan  6 01:41:20 ip-100-64-32-70 kernel: [<ffffffff8128bbee>] ? file_has_perm+0xae/0xc0
Jan  6 01:41:20 ip-100-64-32-70 kernel: [<ffffffff81640d01>] ? __do_page_fault+0xb1/0x450
Jan  6 01:41:20 ip-100-64-32-70 kernel: [<ffffffff811f20f1>] SyS_ioctl+0xa1/0xc0
Jan  6 01:41:20 ip-100-64-32-70 kernel: [<ffffffff816458c9>] system_call_fastpath+0x16/0x1b
```

**Additional information you deem important (e.g. issue happens only occasionally):**

More likely to happen under container load. Happens more often without `MountFlags=slave`.

**Output of `docker version`:**

```
Client:
 Version:      1.12.3
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   6b644ec
 Built:        
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.3
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   6b644ec
 Built:        
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 109
 Running: 36
 Paused: 0
 Stopped: 73
Images: 20
Server Version: 1.12.3
Storage Driver: devicemapper
 Pool Name: docker-253:0-67160128-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 5.369 GB
 Backing Filesystem: xfs
 Data file: /dev/docker/data
 Metadata file: /dev/docker/metadata
 Data Space Used: 18.52 GB
 Data Space Total: 456.3 GB
 Data Space Available: 437.8 GB
 Metadata Space Used: 30.58 MB
 Metadata Space Total: 17.05 GB
 Metadata Space Available: 17.02 GB
 Thin Pool Minimum Free Space: 45.63 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: true
 Deferred Deletion Enabled: true
 Deferred Deleted Device Count: 0
 Library Version: 1.02.135-RHEL7 (2016-09-28)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: overlay null host bridge
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 3.10.0-327.4.5.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 32
Total Memory: 239.8 GiB
Name: xxx
ID: ESAP:AL7B:RXQI:IT2Y:RSFH:3RTK:A7MI:LGGM:BEMK:KBVM:NKXD:ZVWQ
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

AWS

**Comment by cpuguy83** [comment]
Thanks for opening a fresh issue.
Looks like devicemapper is hanging:

```
github.com/docker/docker/pkg/devicemapper._C2func_dm_task_run(0x7fd3a40231b0, 0x7fd300000000, 0x0, 0x0)
	??:0 +0x4c
github.com/docker/docker/pkg/devicemapper.dmTaskRunFct(0x7fd3a40231b0, 0xc821a91620)
	/root/rpmbuild/BUILD/docker-engine/.gopath/src/github.com/docker/docker/pkg/devicemapper/devmapper_wrapper.go:96 +0x75
github.com/docker/docker/pkg/devicemapper.(*Task).run(0xc820345838, 0x0, 0x0)
	/root/rpmbuild/BUILD/docker-engine/.gopath/src/github.com/docker/docker/pkg/devicemapper/devmapper.go:155 +0x37
github.com/docker/docker/pkg/devicemapper.SuspendDevice(0xc8219c8600, 0x5a, 0x0, 0x0)
	/root/rpmbuild/BUILD/docker-engine/.gopath/src/github.com/docker/docker/pkg/devicemapper/devmapper.go:648 +0x99
github.com/docker/docker/pkg/devicemapper.CreateSnapDevice(0xc821a915f0, 0x25, 0x21, 0xc8219c8600, 0x5a, 0x1f, 0x0, 0x0)
	/root/rpmbuild/BUILD/docker-engine/.gopath/src/github.com/docker/docker/pkg/devicemapper/devmapper.go:780 +0x92
github.com/docker/docker/daemon/graphdriver/devmapper.(*DeviceSet).createRegisterSnapDevice(0xc820433040, 0xc821084080, 0x40, 0xc8210842c0, 0x140000000, 0x0, 0x0)
	/root/rpmbuild/BUILD/docker-engine/.gopath/src/github.com/docker/docker/daemon/graphdriver/devmapper/deviceset.go:861 +0x550
github.com/docker/docker/daemon/graphdriver/devmapper.(*DeviceSet).AddDevice(0xc820433040, 0xc821084080, 0x40, 0xc82025b5e0, 0x45, 0x0, 0x0, 0x0)
	/root/rpmbuild/BUILD/docker-engine/.gopath/src/github.com/docker/docker/daemon/graphdriver/devmapper/deviceset.go:1887 +0xa5c
github.com/docker/docker/daemon/graphdriver/devmapper.(*Driver).Create(0xc82036af00, 0xc821084080, 0x40, 0xc82025b5e0, 0x45, 0x0, 0x0, 0x0, 0x0, 0x0)
	/root/rpmbuild/BUILD/docker-engine/.gopath/src/github.com/docker/docker/daemon/graphdriver/devmapper/driver.go:131 +0x6f
github.com/docker/docker/daemon/graphdriver/devmapper.(*Driver).CreateReadWrite(0xc82036af00, 0xc821084080, 0x40, 0xc82025b5e0, 0x45, 0x0, 0x0, 0x0, 0x0, 0x0)
	/root/rpmbuild/BUILD/docker-engine/.gopath/src/github.com/docker/docker/daemon/graphdriver/devmapper/driver.go:126 +0x86
github.com/docker/docker/daemon/graphdriver.(*NaiveDiffDriver).CreateReadWrite(0xc82021c500, 0xc821084080, 0x40, 0xc82025b5e0, 0x45, 0x0, 0x0, 0x0, 0x0, 0x0)
	<autogenerated>:28 +0xbe
github.com/docker/docker/layer.(*layerStore).CreateRWLayer(0xc82021c580, 0xc82154f980, 0x40, 0xc82025b2c0, 0x47, 0x0, 0x0, 0xc820981380, 0x0, 0x0, ...)
	/root/rpmbuild/BUILD/docker-engine/.gopath/src/github.com/docker/docker/layer/layer_store.go:476 +0x5a9
github.com/docker/docker/daemon.(*Daemon).setRWLayer(0xc820432ea0, 0xc8216d12c0, 0x0, 0x0)
```

**Comment by tonistiigi** [comment]
cc @coolljt0725 @rhvgoyal

**Comment by Calpicow** [comment]
Added note about `MountFlags=slave`

**Comment by rhvgoyal** [comment]
I think this is a known kernel issue and it has been fixed. Trying updating to latest kernel.

Following is upstream fix for this.

commit 2a708cff93f1845b9239bc7d6310aef54e716c6a
Author: Junichi Nomura <j-nomura@ce.jp.nec.com>
Date:   Thu Oct 1 08:31:51 2015 +0000

    dm: fix AB-BA deadlock in __dm_destroy()

**Comment by olmoser** [comment]
I'm also facing a stalled docker daemon, getting a similar error message, but it's not the same:

```
[38043.938998] INFO: task dockerd-current:21812 blocked for more than 120 seconds.
[38043.939519] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[38043.940040] dockerd-current D 0000000000000000     0 21812      1 0x00000080
[38043.940625]  ffff883ffcc5b9f8 0000000000000082 ffff883ffcf15e20 ffff883ffcc5bfd8
[38043.941165]  ffff883ffcc5bfd8 ffff883ffcc5bfd8 ffff883ffcf15e20 ffff881ffd069800
[38043.941764]  ffff881ffd069878 0000000000071264 0000000000000000 0000000000000000
[38043.942318] Call Trace:
[38043.942871]  [<ffffffff8168b879>] schedule+0x29/0x70
[38043.943424]  [<ffffffffa007a085>] wait_transaction_locked+0x85/0xd0 [jbd2]
[38043.943939]  [<ffffffff810b1720>] ? wake_up_atomic_t+0x30/0x30
[38043.944501]  [<ffffffffa007a368>] add_transaction_credits+0x268/0x2f0 [jbd2]
[38043.945024]  [<ffffffff810d0c18>] ? enqueue_task_fair+0x208/0x6c0
[38043.945624]  [<ffffffff810c7e55>] ? sched_clock_cpu+0x85/0xc0
[38043.946175]  [<ffffffffa007a5e1>] start_this_handle+0x1a1/0x430 [jbd2]
[38043.946783]  [<ffffffff810c4e18>] ? try_to_wake_up+0x1c8/0x320
[38043.947363]  [<ffffffff811de24a>] ? kmem_cache_alloc+0x1ba/0x1e0
[38043.947938]  [<ffffffffa007aa63>] jbd2__journal_start+0xf3/0x1e0 [jbd2]
[38043.948499]  [<ffffffffa010399a>] ? ext4_dirty_inode+0x2a/0x60 [ext4]
[38043.949019]  [<ffffffffa0131ad9>] __ext4_journal_start_sb+0x69/0xe0 [ext4]
[38043.949576]  [<ffffffffa010399a>] ext4_dirty_inode+0x2a/0x60 [ext4]
[38043.950097]  [<ffffffff81228f0a>] __mark_inode_dirty+0xca/0x290
[38043.950714]  [<ffffffff81219ad1>] update_time+0x81/0xd0
[38043.951258]  [<ffffffff81219d20>] file_update_time+0xa0/0xf0
[38043.951818]  [<ffffffff81182bd8>] __generic_file_aio_write+0x198/0x400
[38043.952362]  [<ffffffff81182e99>] generic_file_aio_write+0x59/0xa0
[38043.952897]  [<ffffffffa00f6f0b>] ext4_file_write+0xdb/0x470 [ext4]
[38043.953406]  [<ffffffff811806ab>] ? unlock_page+0x2b/0x30
[38043.953936]  [<ffffffff811aed8b>] ? do_wp_page+0x17b/0x530
[38043.954451]  [<ffffffff810c4fc0>] ? wake_up_state+0x10/0x20
[38043.954958]  [<ffffffff811fda0d>] do_sync_write+0x8d/0xd0
[38043.955455]  [<ffffffff811fe27d>] vfs_write+0xbd/0x1e0
[38043.955980]  [<ffffffff811fed9f>] SyS_write+0x7f/0xe0
[38043.956483]  [<ffffffff816967c9>] system_call_fastpath+0x16/0x1b
```

I'm also seeing a lot of 

```
device-mapper: thin: Deletion of thin device nnn failed.
```

messages.

We are running RHEL 7.3 with the latest kernel (`3.10.0-514.6.1.el7.x86_64`) which apparently has the deadlock issue mentioned by @rhvgoyal fixed.

**kernel version**
```3.10.0-514.6.1.el7.x86_64```

**docker version**
```
Client:
 Version:         1.12.5
 API version:     1.24
 Package version: docker-common-1.12.5-14.el7.x86_64
 Go version:      go1.7.4
 Git commit:      047e51b/1.12.5
 Built:           Wed Jan 11 17:53:20 2017
 OS/Arch:         linux/amd64

Server:
 Version:         1.12.5
 API version:     1.24
 Package version: docker-common-1.12.5-14.el7.x86_64
 Go version:      go1.7.4
 Git commit:      047e51b/1.12.5
 Built:           Wed Jan 11 17:53:20 2017
 OS/Arch:         linux/amd64
```

**docker info**
```
Containers: 193
 Running: 40
 Paused: 0
 Stopped: 153
Images: 53
Server Version: 1.12.5
Storage Driver: devicemapper
 Pool Name: vg_sys-docker--pool
 Pool Blocksize: 524.3 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file:
 Metadata file:
 Data Space Used: 27.45 GB
 Data Space Total: 79.3 GB
 Data Space Available: 51.85 GB
 Metadata Space Used: 5.12 MB
 Metadata Space Total: 302 MB
 Metadata Space Available: 296.9 MB
 Thin Pool Minimum Free Space: 7.929 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: true
 Deferred Deletion Enabled: true
 Deferred Deleted Device Count: 7
 Library Version: 1.02.135-RHEL7 (2016-11-16)
Logging Driver: json-file
Cgroup Driver: systemd
Plugins:
 Volume: local
 Network: overlay host bridge null
 Authorization: rhel-push-plugin
Swarm: inactive
Runtimes: docker-runc runc
Default Runtime: docker-runc
Security Options: seccomp
Kernel Version: 3.10.0-514.6.1.el7.x86_64
Operating System: Red Hat Enterprise Linux Server 7.3 (Maipo)
OSType: linux
Architecture: x86_64
Number of Docker Hooks: 2
CPUs: 56
Total Memory: 251.6 GiB
Name: lxk8sop001.at.inside
ID: ANEN:5EQJ:ULOK:BRFE:LKDW:MDON:I5LH:DBYY:IG47:F5SB:WEZR:BFTF
Docker Root Dir: /data/docker
Debug Mode (client): false
Debug Mode (server): false
Http Proxy: http://10.3.133.10:8080
Registry: https://index.docker.io/v1/
Insecure Registries:
 127.0.0.0/8
Registries: docker.io (secure)
```

**docker flags**
```
/usr/bin/dockerd-current --add-runtime docker-runc=/usr/libexec/docker/docker-runc-current --default-runtime=docker-runc --authorization-plugin=rhel-push-plugin --exec-opt native.cgroupdriver=systemd --userland-proxy-path=/usr/libexec/docker/docker-proxy-current --log-level=warn --log-opt max-size=2G --log-opt max-file=1 --storage-opt dm.fs=xfs --storage-driver=devicemapper --storage-opt=dm.thinpooldev=/dev/mapper/vg_sys-docker--pool --storage-opt=dm.use_deferred_removal=true --storage-opt=dm.use_deferred_deletion=true
```

**systemd unit**
```
[Unit]
Description=Docker Application Container Engine
Documentation=http://docs.docker.com
After=network.target
Wants=docker-storage-setup.service
Requires=rhel-push-plugin.socket

[Service]
Type=notify
NotifyAccess=all
EnvironmentFile=-/etc/sysconfig/docker
EnvironmentFile=-/etc/sysconfig/docker-storage
EnvironmentFile=-/etc/sysconfig/docker-network
Environment=GOTRACEBACK=crash
Environment=DOCKER_HTTP_HOST_COMPAT=1
Environment=PATH=/usr/libexec/docker:/usr/bin:/usr/sbin
ExecStart=/usr/bin/dockerd-current \
          --add-runtime docker-runc=/usr/libexec/docker/docker-runc-current \
          --default-runtime=docker-runc \
          --authorization-plugin=rhel-push-plugin \
          --exec-opt native.cgroupdriver=systemd \
          --userland-proxy-path=/usr/libexec/docker/docker-proxy-current \
          $OPTIONS \
          $DOCKER_STORAGE_OPTIONS \
          $DOCKER_NETWORK_OPTIONS \
          $ADD_REGISTRY \
          $BLOCK_REGISTRY \
          $INSECURE_REGISTRY
ExecReload=/bin/kill -s HUP $MAINPID
LimitNOFILE=1048576
LimitNPROC=1048576
LimitCORE=infinity
TimeoutStartSec=0
Restart=on-abnormal
MountFlags=slave

[Install]
WantedBy=multi-user.target
```

**dmsetup ls**
```
docker-253:9-20447236-74b14da0413edb9f19ce01285f25f39c10c619e0989fe06ed3ddae1bc32319d0  (253:49)
docker-253:9-20447236-d2ff30ef57e154134a773aa7b27e4c2460dc473a2ac315d259f5a91d06774a35  (253:31)
docker-253:9-20447236-fff1506d2f9e218185225f65c61b8ad26cb5616ab120a81be6ea7a7cbb692a99  (253:105)
docker-253:9-20447236-a0150e2d7e0629e300b76669c28f823b7c4b863e74aaa9588e1816b5fbfe6d26  (253:12)
docker-253:9-20447236-fa95b603891865b88f3af09893af8a7b1014e196271968f0a0078a0df6ffdfcf  (253:65)
docker-253:9-20447236-093914aeec4412cd7730a16b3c50a5919b3d8dd9513ee1acec9ebf71c7a46449  (253:41)
docker-253:9-20447236-97adaab23a0f88d2b680d8fb9300bc84ac684a4206853bd1adba38eb4bf74e14  (253:14)
docker-253:9-20447236-ec70c2c688d16ebb0ab0e32112baaec36411e7d10df5cad9dad1b64c3cf04aac  (253:76)
vg_app-lv_data  (253:9)
docker-253:9-20447236-95b1c467ad5d042d05d774316fd146b1ccdd4a15a078d6975b67e7f87abdeb8c  (253:34)
docker-253:9-20447236-a9d578c6c7a885a97150f42feda2e97b683a0691c258a353cbbf859cb6e3cd16  (253:26)
docker-253:9-20447236-38c70393f9f1540179554851f30fcaf723eabda597c97dee7711511ab1785a9e  (253:81)
docker-253:9-20447236-9484b543a9f1470ba12975edeb8c22cf339281ec63b757c117e9335181aa8de4  (253:23)
docker-253:9-20447236-3cbdabd8aafa48e5a33b240e6c539c62e80ef21b0eaa433cca5c5075f1dd9297  (253:71)
docker-253:9-20447236-f08fc5af478ae42ce59aedbb6718569e68d1a1cf288f8572d61bf27185ae0f8f  (253:44)
docker-253:9-20447236-9eccfd8455a9111213d0157845497b5ef049dae8a58cd666bec086dcf9d53505  (253:30)
vg_sys-tmp      (253:2)
docker-253:9-20447236-b79863122b30bcd4c6123401393a5fe6103633b68b519335bc275f88071d417c  (253:56)
docker-253:9-20447236-57c9ba6dc2132db84d76c367dadd2b12368673cd1b244281bde2012c65198bb5  (253:100)
docker-253:9-20447236-1e88bbe8a2e1d9471c1eee2be483731be5bf612158b4402209218d70b1006bee  (253:103)
docker-253:9-20447236-8d1cfa7077603cb2b412223dd1c4970af917d47821bef9c3ab3b2bebe50e7c1f  (253:85)
docker-253:9-20447236-c82b46a7850d92e5e9b0363a68d550eeb2b961534af1ea1a30405b3a2b881587  (253:63)
docker-253:9-20447236-f00f7876a0228af0bdafee0472719823dea06a4cc291b692d9369034e2b7f48c  (253:62)
docker-253:9-20447236-e1738bfe0a2d434ec8f7969a30ee59258274fe8bb3230228582579616fefb4ea  (253:55)
docker-253:9-20447236-1e40f458b296306b5133d25eb02b615a8afa15f6dbdcf8978f8e9fb6f4c75b09  (253:29)
docker-253:9-20447236-56f97ab6effed7b12e5d109cdbe6cc52c359110b1c4c6858c950a49cdd68152f  (253:15)
vg_sys-swap     (253:1)
docker-253:9-20447236-348a5c7e6f51c04e83700efb7c5deb61a2aabd99e867e5738d5fe19af579a909  (253:61)
docker-253:9-20447236-e631e67c5afeff65be96986802fb95da8decb10b5eb4310476bdcc127fa3b4b6  (253:20)
vg_sys-root     (253:0)
docker-253:9-20447236-4838bd74fb89a400bd860dda4dac8e368f6894e6fe04dfab57ed0096b46cd7ae  (253:77)
docker-253:9-20447236-c993ec29b34137155935d4d24bc7742415cdff6e9f038575a0cf24534dcd037c  (253:57)
docker-253:9-20447236-d27ff498e493c0ce7f4983d3d12f1d68cbb299242bc5ae2d5c8b1c39cce09b23  (253:106)
docker-253:9-20447236-754980e6f7b4874f9b2aaf66f9390349e8e9447270469aeb3f0abd2677958b0d  (253:94)
docker-253:9-20447236-044c721a3993099ffa941bf7547a842fda777a34984decf5f3fcb0f13edacead  (253:59)
docker-253:9-20447236-95976887656dab9bd6a33a962ecc080dc5dc8398a4b5215576537f01e962f298  (253:37)
docker-253:9-20447236-d2d7d9fafa26392590c7b64f0c4ae941eaa2af84688f00e29f83d9f5374d7d7c  (253:72)
docker-253:9-20447236-20f55772419af85ae0db11f9f83f971aff50e3ae70773d12eb62698e7bda58cb  (253:22)
vg_sys-docker--pool_tdata       (253:8)
docker-253:9-20447236-ca7bece5f60d6573e4baac64c3174ce5fd9a4e11444c427c403d3ab29b27006f  (253:53)
docker-253:9-20447236-d3f75f77bb2917fbf8bc890bff0f0c777e93034f096e2e26506afef9f6a126eb  (253:42)
docker-253:9-20447236-789ed39325145bac29efbea1ddf2c880ea9ddb86d479d52f839e77484f9cfac5  (253:99)
docker-253:9-20447236-7013dc8ccd581b5e47485d6a96e263950393a1cebb1f069e0d1f9e289d3ee5b3  (253:90)
docker-253:9-20447236-fb58c58118a7d8bf1bd9723c1d3434b825dd79d6be6e8fc161232310e10d293d  (253:95)
docker-253:9-20447236-94db7ac318882d9a50e2adb6dfadd881c38dbee821faffaf9db3a736d58244f4  (253:87)
docker-253:9-20447236-e2efb64595fa6a96c7b7ce763e149eb85f1b3d25c2b9f4841a047e3cb86852ae  (253:67)
vg_sys-docker--pool_tmeta       (253:7)
docker-253:9-20447236-aed65478e456dfd547e0d76dc26dd62b6e6305b29f12363790095d405e255a93  (253:38)
docker-253:9-20447236-1efeac4bb9df0343874cbc824a4837e221dd95d5c7254c0cc6e20fb440a77274  (253:21)
vg_sys-var      (253:4)
docker-253:9-20447236-4a0fff450d417dad1e82b1fb70ae709da36fc8de04a48639144f40dded29a3fe  (253:79)
docker-253:9-20447236-8e766abc6a8cb76d032df35202afd82b48187b011afbf41067798df70ecf8729  (253:104)
docker-253:9-20447236-5e79d838de505393ed46410bf2eea9b2d8e4f9fb7e4c8e2d519215da53807744  (253:36)
docker-253:9-20447236-70359c38e4cf9ca29e959458af36943f895f715c5cb51b6fd726406689e10f47  (253:96)
docker-253:9-20447236-e4c0485a6819dce7ba819c6db5a22587cbdf6407e1797d2eb1a7f3c4fdcc7c30  (253:75)
docker-253:9-20447236-b4d2e1b1aadb1dcbe7b100ca477243b9892680ce353be5fec3afa010b6516c8f  (253:58)
docker-253:9-20447236-4c3b946fc078c1312a588df0a726e9c4c66117d6fc7f9fdde7ca1d03acff4db4  (253:13)
docker-253:9-20447236-d989a5114bc3e6709ce6ede12c11fffce389fa74deb4ce8ebca2004a752bce08  (253:54)
docker-253:9-20447236-dd3f40cd7060d505ecc212264b99ad24e27e71ebc5a0baba7e2d4fffab381c70  (253:18)
docker-253:9-20447236-df88f1e2dab7523b1fd19ccf56948ffb836a865f0f03723ea086750f2c094978  (253:35)
docker-253:9-20447236-aeb53af120818e2c9e66b31c7be2ce724587ec58571e0109776d9282fd6ed547  (253:52)
docker-253:9-20447236-02a054fddbd77889e4ed9c940a1e57eb6bb6bfce6a9e35131e00046d5636e7df  (253:102)
docker-253:9-20447236-8f253a7efbe0455359f9bcc1a73d68a35cc1bc8971673750bb739588ce8a50cd  (253:48)
docker-253:9-20447236-41a2d3a8bfec862019c951ef90ec62cb03b9c7ec343ccf85c1cfa450edd652d8  (253:46)
docker-253:9-20447236-670a2a4edb50a35188404474feeb90f2be56de905db23a28a01a7c331066f1d5  (253:84)
docker-253:9-20447236-a1a867451f5b5b4b326dbe3dec791448022332e5cd0616e6f02a35ded56cbefd  (253:78)
docker-253:9-20447236-59aae433b17a7005f5973b50faee19c6e6a7d43aabffdf6feb679fa3374a02b1  (253:40)
docker-253:9-20447236-d6759dbdaeb0cc276bf32eade16dbb2a54e04a6bb618a31e7ac4e2f60ca76a4e  (253:92)
docker-253:9-20447236-258355baf6b341f3fc95fddbeecef7b82dd42cfe2bca302a19bd4b3ee05a1cea  (253:89)
docker-253:9-20447236-e1539129708c380ecdc1c4cc859b2c437608632c6095352a73cb83efb2f6c15c  (253:107)
docker-253:9-20447236-1ef5ad8f1b46126191bcd9857dae456446ba0a23af94653ed4a0db2800603e8c  (253:97)
docker-253:9-20447236-9c022fb249a1bc5eaf3b2b11c27e5ceacb85d1b6b6a97b9e9e1f64089f27a2fb  (253:108)
docker-253:9-20447236-a44b881ef39ee030d37b29d2de298696661bf22214eb97e2fe0ec3e4208a95cd  (253:11)
docker-253:9-20447236-48e5c8ee8919c56be04ccddab926e7e533537fcaeb0c1e1314f30b4960cf04d8  (253:68)
docker-253:9-20447236-95fc4d254b5038bd98a2713018ff788819917444539b37f6dcdd4a725a23ce6b  (253:86)
docker-253:9-20447236-dbb9443c4952653ce00c1de4d69d955cbcd01db57b575216a13efb4d8376455b  (253:74)
docker-253:9-20447236-19e774a9b94793a8252f2001e2c22123d5adb995c5acd566ad4fb33bb575bfb9  (253:27)
docker-253:9-20447236-00c168f4803b8d039095303837a15ce9ea1914cb02746a212d2bf581f5b70ea8  (253:25)
docker-253:9-20447236-cda8774711c76249fc095f10aa060175c6bab3c4daf3b8b026c09f4eec3f0428  (253:47)
docker-253:9-20447236-a778b09881b68ad1279ec22017d48a1cdb4c71eab8272662bd4d2e842ab4ce10  (253:39)
docker-253:9-20447236-fed49c0ec6bb464f46b1d9238c9f3b998a46ea6cb66b1317c329b924ec9e0828  (253:33)
docker-253:9-20447236-83b774920c8f11fe94b11407ab6ae85e9c01b1a530cccc275c5bf338b79974b0  (253:64)
docker-253:9-20447236-6752d385bf3fcba896c480091223ca3125803be2754a294ab8cab3f44035c423  (253:45)
docker-253:9-20447236-58cc65055ba7879cc2309f13747fc77c07ea129fa9c50a32c17262a3457b8cfc  (253:32)
docker-253:9-20447236-5f72ea9d38128813ef1bf927cc66e6ff4719e244acf8b77d66ec4ac80481839a  (253:19)
vg_sys-var_log  (253:5)
docker-253:9-20447236-6a2e6d3508de09005dd8398dc8417c25458984156e00d7e9b09eaf6c1848ba7f  (253:80)
vg_sys-opt      (253:6)
docker-253:9-20447236-07b688dec2f5b5cf2e2c66f591c03a64bf69cdd6b1588e9fc559e39fbc8748d7  (253:16)
docker-253:9-20447236-97d19a0b31b9dcfd8e5d1c08207ff275519dba455340f84f3ff520673e481288  (253:88)
docker-253:9-20447236-e0a75318d89f4a2fd63c9d270e8270c7dd69bf93e095c8242c458b356809dda4  (253:60)
docker-253:9-20447236-b4af3e18e8f68a61b3fbf0bf32eea06b967907c42f3046a98d70609b01934976  (253:28)
docker-253:9-20447236-5ae28d7e353b09caf37e5a67d52f48c402e1626ecf248d89445b72ac2ef8a963  (253:66)
docker-253:9-20447236-3ab06ee53afd240354d6ee5939f525c7b66a814d88bad6854dee1957d82d5930  (253:17)
docker-253:9-20447236-76c992a5b15725cbf99276227527dc321ad612386ff61d7b25be883cc0564b8e  (253:93)
docker-253:9-20447236-aff63934136457bf34ab57a0cf98373b32cdd91c6280e0a78feb9ab1216a6f0d  (253:73)
docker-253:9-20447236-4fa2185fedc8babc5507d5c5fad8805e985ae2ac2e8ac4693e9a47254e63aec4  (253:101)
vg_sys-docker--pool     (253:10)
docker-253:9-20447236-bdb746322182ea1f2b2098b364c2b00c505c63b49444aebb5062fad591b053c2  (253:98)
docker-253:9-20447236-25107532ca08b90f0a7d852900d680a3dc7a4409f4cb9bc3fb40431eec71e654  (253:24)
docker-253:9-20447236-8f4823cece161fbf1f50ec9fe26054561614fe75849b7268753f818d0c561fe6  (253:91)
docker-253:9-20447236-a845d09c24566c121a17c23c11f84def42ad931ba3bf1cf1fc12d8292edecda7  (253:50)
docker-253:9-20447236-bff41caaa3718d9e0c1734868b54fee86a7986dad5ec98a3e1ffaf95b47d043e  (253:83)
docker-253:9-20447236-4ff47a9ca1c23460edda88dfc68ce2ec5b6886943dd8434ab4e836a5bdb42b0c  (253:70)
docker-253:9-20447236-cb6e42dce826191247a1de32fc94d4a9461a1e8a6340493f16b810c7bdba0a4c  (253:51)
docker-253:9-20447236-c3a71dc11c46b973eb956c7496f12c66b8c3e8e74335ad391b792be94a8b3733  (253:43)
docker-253:9-20447236-4f9b4109859e94a018bdb259a83fdb6220be5361a80ddca8cedf4bb6dd2e53d9  (253:82)
docker-253:9-20447236-9ef70894a10f7e11d44022dab9ad683affd8cc43e00310bc0a6a26a3a90e7e17  (253:69)
vg_sys-home     (253:3)
```

any suggestions highly appreciated, this is a big pain for us in production since dockerd is defunct at this point and only a reboot resolves the situation.

**Comment by fsfshijm** [comment]
+1, the same issue
```
[3024480.475038] INFO: task loop0:22588 blocked for more than 120 seconds.
[3024480.475650] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[3024480.476400] loop0 D ffff88043fc93680 0 22588 2 0x00000080
[3024480.476403] ffff8803f992b070 0000000000000046 ffff880426b4f1c0 ffff8803f992bfd8
[3024480.476405] ffff8803f992bfd8 ffff8803f992bfd8 ffff880426b4f1c0 ffff88027da538c8
[3024480.476407] ffff88027da538d0 7fffffffffffffff ffff880426b4f1c0 0000000000000000
[3024480.476409] Call Trace:
[3024480.476412] [<ffffffff81609259>] schedule+0x29/0x70
[3024480.476415] [<ffffffff816071a9>] schedule_timeout+0x209/0x2d0
[3024480.476417] [<ffffffff812ae924>] ? blk_finish_plug+0x14/0x40
[3024480.476426] [<ffffffffa041920c>] ? _xfs_buf_ioapply+0x31c/0x420 [xfs]
[3024480.476429] [<ffffffff81609756>] wait_for_completion+0x116/0x170
[3024480.476431] [<ffffffff810a94d0>] ? wake_up_state+0x20/0x20
[3024480.476438] [<ffffffffa041b11c>] ? xfs_bwrite+0x2c/0x60 [xfs]
[3024480.476446] [<ffffffffa041ada3>] xfs_buf_iowait+0x33/0xc0 [xfs]
[3024480.476453] [<ffffffffa041b11c>] xfs_bwrite+0x2c/0x60 [xfs]
[3024480.476461] [<ffffffffa0422d1b>] xfs_reclaim_inode+0x32b/0x350 [xfs]
[3024480.476469] [<ffffffffa0422fa7>] xfs_reclaim_inodes_ag+0x267/0x390 [xfs]
[3024480.476472] [<ffffffff811dc375>] ? dentry_lru_del+0x65/0x70
[3024480.476474] [<ffffffff811de7c4>] ? shrink_dentry_list+0x64/0x480
[3024480.476482] [<ffffffffa0423b23>] xfs_reclaim_inodes_nr+0x33/0x40 [xfs]
[3024480.476491] [<ffffffffa042e7d5>] xfs_fs_free_cached_objects+0x15/0x20 [xfs]
[3024480.476493] [<ffffffff811c9e26>] prune_super+0xf6/0x1a0
[3024480.476495] [<ffffffff81168de5>] shrink_slab+0x165/0x300
[3024480.476497] [<ffffffff811c00d1>] ? vmpressure+0x21/0x90
[3024480.476499] [<ffffffff8116bf32>] do_try_to_free_pages+0x3c2/0x4e0
[3024480.476501] [<ffffffff8116c14c>] try_to_free_pages+0xfc/0x180
[3024480.476505] [<ffffffff811607cd>] __alloc_pages_nodemask+0x7fd/0xb90
[3024480.476508] [<ffffffff8119efd9>] alloc_pages_current+0xa9/0x170
[3024480.476510] [<ffffffff811a9035>] new_slab+0x275/0x300
[3024480.476513] [<ffffffff81600e9b>] __slab_alloc+0x315/0x48f
[3024480.476517] [<ffffffff812d9325>] ? __radix_tree_preload+0x35/0xa0
[3024480.476519] [<ffffffff811bb832>] ? __mem_cgroup_commit_charge+0x152/0x390
[3024480.476521] [<ffffffff811ab673>] kmem_cache_alloc+0x193/0x1d0
[3024480.476523] [<ffffffff812d9325>] ? __radix_tree_preload+0x35/0xa0
[3024480.476525] [<ffffffff812d9325>] __radix_tree_preload+0x35/0xa0
[3024480.476527] [<ffffffff812d93e9>] radix_tree_maybe_preload+0x19/0x20
[3024480.476529] [<ffffffff81156e0b>] __add_to_page_cache_locked+0x8b/0x260
[3024480.476531] [<ffffffff81157037>] add_to_page_cache_lru+0x37/0xb0
[3024480.476533] [<ffffffff8115748f>] grab_cache_page_write_begin+0x8f/0xd0
[3024480.476560] [<ffffffffa018c667>] ext4_da_write_begin+0xa7/0x340 [ext4]
[3024480.476562] [<ffffffff8115648d>] generic_file_buffered_write+0x11d/0x290
[3024480.476565] [<ffffffff811585f5>] __generic_file_aio_write+0x1d5/0x3e0
[3024480.476567] [<ffffffff8115885d>] generic_file_aio_write+0x5d/0xc0
[3024480.476572] [<ffffffffa0181b75>] ext4_file_write+0xb5/0x460 [ext4]
[3024480.476575] [<ffffffff81600814>] ? __slab_free+0x10e/0x277
[3024480.476578] [<ffffffff811c5e3d>] do_sync_write+0x8d/0xd0
[3024480.476582] [<ffffffffa03557bf>] __do_lo_send_write+0x6f/0x120 [loop]
[3024480.476585] [<ffffffffa03558d2>] do_lo_send_direct_write+0x62/0x80 [loop]
[3024480.476587] [<ffffffffa0355c07>] do_bio_filebacked+0x257/0x320 [loop]
[3024480.476590] [<ffffffffa0355870>] ? __do_lo_send_write+0x120/0x120 [loop]
[3024480.476596] [<ffffffffa0333748>] ? clone_endio+0xc8/0xf0 [dm_mod]
[3024480.476598] [<ffffffffa035643c>] loop_thread+0xfc/0x300 [loop]
[3024480.476601] [<ffffffff81098230>] ? wake_up_bit+0x30/0x30
[3024480.476603] [<ffffffffa0356340>] ? loop_attr_do_show_backing_file+0xe0/0xe0 [loop]
[3024480.476605] [<ffffffff8109726f>] kthread+0xcf/0xe0
[3024480.476607] [<ffffffff810971a0>] ? kthread_create_on_node+0x140/0x140
[3024480.476610] [<ffffffff81613cfc>] ret_from_fork+0x7c/0xb0
[3024480.476611] [<ffffffff810971a0>] ? kthread_create_on_node+0x140/0x140
```

**Comment by fsfshijm** [comment]
@olmoser seems the following content can fix the issue of 'hung_task_timeout_secs'
https://www.blackmoreops.com/2014/09/22/linux-kernel-panic-issue-fix-hung_task_timeout_secs-blocked-120-seconds-problem/
https://lonesysadmin.net/2013/12/22/better-linux-disk-caching-performance-vm-dirty_ratio/

**Comment by iclouding** [comment]
```
sudo sysctl -w vm.dirty_ratio=80
sudo sysctl -w vm.dirty_background_ratio=5
sudo sysctl -p
```

@fsfshijm , I have encountered the same issue, can it works?

**Comment by gswallow** [comment]
Is there any hope of a fix for this seeing light of day?  I just installed docker-engine 17.04.0 and now docker ps, docker info, ALL docker commands are stalling.

**Comment by cpuguy83** [comment]
@gswallow Doesn't mean you are hitting this issue.
Can you send `SIGUSR1` to the daemon and collect the goroutine stack dump from the logs?

Thanks.

**Comment by gswallow** [comment]
Where do those get dumped when I send SIGUSR1's?

**Comment by cpuguy83** [comment]
@gswallow In the daemon logs... depends on your host OS/init system.
Likely from `journalctl -uf docker`

**Comment by gswallow** [comment]
Ok:

docker version 1.13.1 and versions prior.  Ubuntu 16.04.2.  Docker info below.  'docker ps' hangs indefinitely.

Load average: 7.02 and it just rises...

docker info
Containers: 82
 Running: 6
 Paused: 0
 Stopped: 76
Images: 25
Server Version: 1.13.1
Storage Driver: devicemapper
 Pool Name: vg0-pool0
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: ext4
 Data file: 
 Metadata file: 
 Data Space Used: 19.77 GB
 Data Space Total: 73.01 GB
 Data Space Available: 53.24 GB
 Metadata Space Used: 34.78 MB
 Metadata Space Total: 4.291 GB
 Metadata Space Available: 4.256 GB
 Thin Pool Minimum Free Space: 7.301 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: true
 Deferred Deletion Enabled: true
 Deferred Deleted Device Count: 0
 Library Version: 1.02.110 (2015-10-30)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: aa8187dbd3b7ad67d8e5e3a15115d3eef43a7ed1
runc version: 9df8b306d01f59d3a8029be411de015b7304dd8f
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-67-generic
Operating System: Ubuntu 16.04.2 LTS
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 14.69 GiB
Name: ip-172-21-205-177
ID: U3MQ:ONOL:YG4A:JLZ2:KVMJ:RN6V:4X76:DIME:6K3O:7TR4:K5WK:ZBOK
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

[goroutine-stacks-2017-04-25T193435Z.txt](https://github.com/moby/moby/files/956337/goroutine-stacks-2017-04-25T193435Z.txt)

**Comment by cpuguy83** [comment]
@gswallow It looks like you've got a number of calls trying to walk the filesystem tree to get the container size.
How big are your containers? Are you running `docker system df`? In parallel?

**Comment by cpuguy83** [comment]
@gswallow In any case, this is very much a different issue, can you open a new one? Thanks.

**Comment by pfremm** [comment]
@Calpicow we maybe having a similar issue.  Out of curiosity are you setting either dm.use_deferred_removal or dm.use_deferred_deletion to true?  Curious if that would help.

**Comment by Calpicow** [comment]
@pfremm Both flags are set to true.

For my particular issue, I could confirm a kernel update resolves it.

**Comment by thaJeztah** [comment]
Let me close this issue as the original issue was a kernel issue (https://github.com/moby/moby/issues/30003#issuecomment-272955488) and resolved, but if you're running into this on an up-to-date system and docker version, please open a new issue to prevent possibly unrelated issues from being conflated, which makes it hard to track 👍

**Comment by pfremm** [comment]
@calpicoh  what kernel version worked?

**Comment by Calpicow** [comment]
@pfremm CentOS 7.2 3.10.0-514.6.1.el7.x86_64
